src/lapack/backends/cusolver/cusolver_batch.cpp

```diff
@@ -79,251 +100,642 @@ void getri_batch(sycl::queue &queue, std::int64_t n, sycl::buffer<std::complex<d
+                           group_count, group_sizes, scratchpad, scratchpad_size, dependencies); \
+    }
+
+GEQRF_BATCH_LAUNCHER_USM(float, cusolverDnSgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(double, cusolverDnDgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(std::complex<float>, cusolverDnCgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(std::complex<double>, cusolverDnZgeqrf)
+
+#undef GEQRF_BATCH_LAUNCHER_USM
+
+template <typename Func, typename T>
+inline sycl::event getrf_batch(const char *func_name, Func func, sycl::queue &queue, std::int64_t m,
+                               std::int64_t n, T *a, std::int64_t lda, std::int64_t stride_a,
+                               std::int64_t *ipiv, std::int64_t stride_ipiv,
+                               std::int64_t batch_size, T *scratchpad, std::int64_t scratchpad_size,
+                               const std::vector<sycl::event> &dependencies) {
+    using cuDataType = typename CudaEquivalentType<T>::Type;
+
+    overflow_check(m, n, lda, stride_a, stride_ipiv, batch_size, scratchpad_size);
+
+    // cuSolver legacy api does not accept 64-bit ints.
+    // To get around the limitation.
+    // Allocate memory with 32-bit ints then copy over results
+    std::uint64_t ipiv_size = stride_ipiv * batch_size;
+    int *ipiv32 = (int *)malloc_device(sizeof(int) * ipiv_size, queue);
+    int *devInfo = (int *)malloc_device(sizeof(int) * batch_size, queue);
+
+    auto done = queue.submit([&](sycl::handler &cgh) {
+        int64_t num_events = dependencies.size();
+        for (int64_t i = 0; i < num_events; i++) {
+            cgh.depends_on(dependencies[i]);
+        }
+        onemkl_cusolver_host_task(cgh, queue, [=](CusolverScopedContextHandler &sc) {
+            auto handle = sc.get_handle(queue);
+            auto a_ = reinterpret_cast<cuDataType *>(a);
+            auto devInfo_ = reinterpret_cast<int *>(devInfo);
+            auto scratchpad_ = reinterpret_cast<cuDataType *>(scratchpad);
+            auto ipiv_ = reinterpret_cast<int *>(ipiv32);
+            cusolverStatus_t err;
+
+            // Uses scratch so sync between each cuSolver call
+            for (int64_t i = 0; i < batch_size; ++i) {
+                CUSOLVER_ERROR_FUNC_T_SYNC(func_name, func, err, handle, m, n, a_ + stride_a * i,
+                                           lda, scratchpad_, ipiv_ + stride_ipiv * i, devInfo_ + i);
+            }
+        });
+    });
+
+    // Copy from 32-bit USM to 64-bit
+    sycl::event done_casting = queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(done);
+        cgh.parallel_for(sycl::range<1>{ ipiv_size },
+                         [=](sycl::id<1> index) { ipiv[index] = ipiv32[index]; });
+    });
+
+    // Enqueue free memory, don't return event as not-neccessary for user to wait for ipiv32 being released
+    queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(done_casting);
+        cgh.host_task([&](sycl::interop_handle ih) { sycl::free(ipiv32, queue); });
```

Comment:
This will lead to sporadic crashes: passing queue by reference makes sense as queue object is not going to be destroyed before host_tasl lambda is finished, but ipiv32 variable, which is automatic stack-allocated, can be destroyed by the time of host_task execution, and contain some other data, so free it will crash. `[=,&queue]`

Reply:
I have reverted this change